client/src/app/core/services/csv-export.service.ts

```diff
@@ -0,0 +1,118 @@
+import { BaseViewModel } from '../../site/base/base-view-model';
+import { Injectable } from '@angular/core';
+import { FileExportService } from './file-export.service';
+
+/**
+ * Handles all incoming and outgoing notify messages via {@link WebsocketService}.
+ */
+@Injectable({
+    providedIn: 'root'
+})
+export class CsvExportService {
+    protected exporter = new FileExportService();
+
+    /**
+     * empty constructor TODO
+     */
+    public constructor() {}
+
+    /**
+     * Saves an array of model data to a CSV.
+     * @param data Array of Model instances to be saved
+     * @param columns Column definitions (model property; optional csv title)
+     * @param filename name of the resulting file
+     * @param options
+     */
+    public saveAsCsv(
+        data: BaseViewModel[],
+        columns: { property: string; title?: string }[],
+        filename: string,
+        { lineSeparator = '\r\n', columnSeparator = ';' }: { lineSeparator?: string; columnSeparator?: string }
+    ): boolean {
+        const all_lines = []; // Array of arrays of entries
+        const used_columns = []; // mapped properties to be included
+
+        // initial array of usable text separators. The first character not used
+        // in any text data or as column separator will be used as text separator
+        let ts_list = ['"', "'", '`', '/', '\\', ';', '.', ','];
+
+        if (lineSeparator === columnSeparator) {
+            throw new Error('TODO: lineseparator and columnseparator must differ from each other');
+        }
+
+        ts_list = this.checkCsvTextSafety(lineSeparator, ts_list);
+        ts_list = this.checkCsvTextSafety(columnSeparator, ts_list);
+
+        // create header data
+        const header = [];
+        for (const column of columns) {
```

Comment:
Here you have to interate through the key-values in the columns object.

Reply:
So, we'll lose the 'label for this column is optional'. If that's fine, I'll continue with this

Reply:
yeah. If we have too much of the same names, you can define the object values as `string|null`, null meaning "take the property value".